File: hw1abc/abc/lowerTriangularMatrix.cpp
//------------------------------------------------------------------------------
// lowerTriangularMatrix.cpp - содержит функции обработки нижней треугольной матрицы
//-----------------------------------------------------------------------------

#include "lowerTriangularMatrix.h"
#include <stdexcept>

//------------------------------------------------------------------------------
// Ввод параметров нижней треугольной матрицы из потока.
void In(lowerTriangularMatrix &ltm, FILE* file) {
    fscanf(file, "%d", &ltm.dimension);
    ltm.ltmElements = new double [(ltm.dimension*(ltm.dimension + 1))/2];
    for (int i = 0; i < (ltm.dimension * (ltm.dimension + 1)) / 2; ++i) {
        fscanf(file, "%lf", &ltm.ltmElements[i]);
    }
}

// Случайный ввод параметров нижней треугольной матрицы.
void InRandom(lowerTriangularMatrix &ltm) {
    ltm.dimension = Random();
    ltm.ltmElements = new double [(ltm.dimension * (ltm.dimension + 1)) / 2];
    for (int i = 0; i < (ltm.dimension * (ltm.dimension + 1)) / 2; ++i) {
        ltm.ltmElements[i] = RandomDouble();
    }
}

//------------------------------------------------------------------------------
// Вывод параметров нижней треугольной матрицы в поток.
void Out(lowerTriangularMatrix &ltm, FILE* file) {
    fprintf(file,"It is lower triangular matrix: dimension = %d. Arithmetic mean = %.2f", ltm.dimension, ArithmeticMean(ltm));
}

//------------------------------------------------------------------------------
// Вычисление среднего арифметического нижней треугольной матрицы.
double ArithmeticMean(lowerTriangularMatrix &ltm) {
    double sum = 0.0;
    for (int i = 0; i < (ltm.dimension * (ltm.dimension + 1)) / 2; ++i) {
        sum += ltm.ltmElements[i];
    }
    return double(sum / (ltm.dimension * ltm.dimension));
}

File: hw3abc/abc3/lowerTriangularMatrix.py
from matrix import Matrix
import random


# ------------------------------------------------------------------------------
# lowerTriangularMatrix.py - содержит описание нижней треугольной матрицы.
# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# Нижняя треугольная матрица.
class LowerTriangularMatrix(Matrix):
    __size__: int
    __numberOfElements__: int
    __ltmElements__: list

    def __init__(self, size: int):
        super().__init__(size)
        self.__numberOfElements__ = self.__size__ * (self.__size__ + 1) // 2
        self.__ltmElements__ = []

    # Ввод параметров нижней треугольной матрицы из файла.
    def read_file(self, line: str):
        self.__ltmElements__ = list(map(float, line.split()))

    # Случайный ввод параметров нижней треугольной матрицы.
    def random_read(self):
        self.__size__ = random.randint(0, 20) + 1
        self.__numberOfElements__ = self.__size__ * (self.__size__ + 1) // 2
        self.__ltmElements__ = []
        for i in range(0, self.__numberOfElements__):
            self.__ltmElements__.append(random.random() * random.randint(0, 100))

    # Вывод параметров нижней треугольной матрицы в файл.
    def write(self) -> str:
        return "It is lower triangular matrix: dimension = " + str(self.__size__) + ". Arithmetic mean = " + str(
            self.arithmeticMean()) + "\n"

    # Вычисление среднего арифметического нижней треугольной матрицы.
    def arithmeticMean(self) -> float:
        summ = 0
        for i in range(0, self.__numberOfElements__):
            summ = summ + self.__ltmElements__[i]

        return summ / (self.__size__ * self.__size__)
